### ISSUE TITLE
Support robust method for comparing fish versions

### ISSUE THREAD

**Issue by abudden** [fish-shell/fish-shell#4419]
There may already be a way to do this, but I haven't managed to find it.

I'm currently using fish 2.4.0 and 2.6.0 (on different machines).  My configuration is common (version controlled) across multiple machines and sometimes there are new features made available in new versions of fish that I want to use but aren't available yet in the fish versions in one distro's package manager.  In this case, it would be useful to have a robust way of comparing versions like python's `sys.hexversion`, so my config could include:

```fish
if testversion 0x02050000
    # Stuff that relies on fish version 2.5 features
end
if testversion 0x03000000
    # Stuff that relies on fish version 3.0 features
end
```

At the moment I've implemented this with:

```fish
function get_hexversion
	set verparts (string split . $version)
	while test (count $verparts) -lt 4
		set verparts $verparts 0
	end
	echo "0x"(printf "%02x" $verparts[1])(printf "%02x" $verparts[2])(printf "%02x" $verparts[3])(printf "%02x" $verparts[4])
end
function testversion -a expectedhex
	set -l expected (printf "%d" $expectedhex)
	set -l actual (printf "%d" (get_hexversion))
	return (test $actual -ge $expected)
end
```

However, it would be nice to have a built-in way of supporting this (although this obviously wouldn't work on existing released versions).  There's probably a much better way of converting (e.g.) "2.6.0" into "0x02060000" than the way I've done it above (list comprehensions?), but this seems to work at the moment at least until I see what the version string looks like in 2.5b1 or whatever...

I guess it could be done at the moment with a `testversion` implementation that parses "2.6.0" directly, but I implemented it this way due to my familiarity with `sys.hexversion` in python.

**Comment by zanchey** [comment]
`sort` might be of some value here:

```fish
if begin; echo minimum_version; echo $FISH_VERSION; end | sort --check=silent --version-sort
    do_a_thing
end
```

**Comment by faho** [comment]
GNU sort, though.

I've actually implemented a `vercmp` like thing in fish:

```fish
function fish_vercmp --description "Compare versions"
    # This is a pure fish version of rpm's "vercmp" utility.
    # It is useful for e.g. making completions dependent on having at least a certain version.
    # It both prints and returns
    # "-1" if the first version is smaller
    # "1" if the first version is greater
    # "0" if both are equal
    #
    # TODO: Should we automatically match the largest version-looking part of the argument?
    #       This would allow us to not always use "string match" when calling this,
    #       since plenty of tools will output $toolname $version, like `git --version`:
    #       "git version 2.11.1"
    if not set -q argv[2]
        echo "Expected two arguments" >&2
        return 2
    end
    for arg in $argv
        # "-" is for release number, common in packages.
        # TODO: This does not handle "bX" beta versions or -debian~15
        # versions
        # It does, however, handle "2.6.0-610-g55ab7a42" - which is a $FISH_VERSION
        # when building from git. The last component causes `test` to error out,
        # which causes versions with the same number of commits since the same tag to
        # compare equal.
        if not string match -qr '^\d+(\.\d+)*(-\d+)?(-g[a-zA-Z0-9]+)?$' -- $arg
            echo "Invalid version $arg" >&2
            return 124
        end
    end
    set -l first (string split "." -- $argv[1] | string split "-")
    set -l second (string split "." -- $argv[2] | string split "-")
    # Missing trailing components are effectively 0
    while test (count $first) -lt (count $second)
        set first $first 0
    end
    while test (count $second) -lt (count $first)
        set second $second 0
    end
    while set -q first[1]
        # Ignore stderr because we could be comparing the last component.
        if test $first[1] -lt $second[1] ^/dev/null
            echo -1
            return 1
        else if test $first[1] -gt $second[1] ^/dev/null
            echo 1
            return 2
        end
        set -e first[1]
        set -e second[1]
    end
    # No components differ, version is the same.
    echo 0
    return 0
end
```

It prints "-1" and returns "1" if the first argument is smaller, prints "1" and returns "2" if it is bigger and "0" if they are equal. (We can't return negative $status)

Use like `if test (fish_vercmp $FISH_VERSION 2.6.0) -ge 0`.

As the comments say, it does not handle alpha or beta versions.


-----


So, I'd be okay with shipping a tool to determine if a fish version is new enough. Personally, I prefer feature testing, but I can see how that doesn't always work.

TBD is what it should look like. I kinda dislike the hex thing and would prefer if it parsed the version string instead of requiring the user to translate it to hex, though. My above function is also a bit awkward to use, which mostly comes down to the fact that it wants to distinguish three cases - smaller, equal, bigger. If it just did "is at least" it would be simpler, but it shouldn't be called `vercmp` anymore.

**Comment by zanchey** [comment]
`sort -V` is supported in FreeBSD (11, possibly earlier), but not macOS.

**Comment by faho** [comment]
>sort -V is supported in FreeBSD (11, possibly earlier), but not macOS.

Neat. And "--check=silent" is "-C" in POSIX, so:

@abudden: If you're not using macOS, `printf '%s\n' $MIN_VERSION $FISH_VERSION | sort -CV` should work, even with betas and git builds and such.

In any case, if we added something to fish (which as I said I'd support), this would only be useful if you never need to know which version from before it was added it is.

I.e. assuming we add it in fish 3.0, you could determine that it's 3.0 or 3.1, but you would only be able to use its absence as an indicator that you're on a fish < 3.0. If we wrote it as a function, you could add it to any fish that has all the required features for it.

I've now continued improving that vercmp-style function, and the newest feature it requires should be `string`, so any fish since 2.3.0 should be supported. It's turned out to be kinda gnarly, though:

```fish
function fish_vercmp --description "Compare versions"
    # This is a pure fish version of rpm's "vercmp" utility.
    # It is useful for e.g. making completions dependent on having at least a certain version.
    # It both prints and returns
    # "-1" if the first version is smaller
    # "1" if the first version is greater
    # "0" if both are equal
    #
    # This handles really rather weird versions like '1.4.0~pre2+git141-g6d40dace6358-1'
    # (an actual debian package version),
    # and also "2.6.0-610-g55ab7a42" - which is a $FISH_VERSION
    # when building from git.
    #
    # Behavior is weird for some (meaningless) comparisons, e.g.
    # Git versions with the same number of commits compare the hash (same as vercmp).
    if not set -q argv[2]
        echo "Expected two arguments" >&2
        return 2
    end
    # Note: We do _not_ verify that the arguments are versions
    # - the definition we accept is very loose anyway.

    # First split on "." and "-" to remove those characters.
    # Then split on every character to allow comparing beta versions and such.
    set -l first (string split "." -- $argv[1] | string split "-" | string split "")
    set -l second (string split "." -- $argv[2] | string split "-" | string split "")
    # Missing trailing components are effectively 0
    # This also makes 2.6a1 smaller than 2.6 - since the latter is interpreted as "2600", the former as "26a1".
    while test (count $first) -lt (count $second)
        set first $first 0
    end
    while test (count $second) -lt (count $first)
        set second $second 0
    end
    while set -q first[1]
        # Simple case - both are the same character, so skip ahead.
        if test $first[1] = $second[1]
            set -e first[1]
            set -e second[1]
            continue
        else if string match -qr '[0-9]' -- $first[1]; and not string match -qr '[0-9]' -- $second[1]
            # First is numeric, second isn't - first is greater.
            # (I.e. 2.6a1 is _smaller_ than 2.6.0)
            echo 1
            return 2
        else if not string match -qr '[0-9]' -- $first[1]; and string match -qr '[0-9]' -- $second[1]
            # Other way around - second is numeric, so first is smaller.
            echo -1
            return 1
        else if test (printf '%d' "'$first[1]" | string join "") -lt (printf '%d' "'$second[1]" | string join "")
            # Both are numeric or both aren't - compare character value.
            # Note that this is decidedly not locale-aware and encoding dependent.
            echo -1
            return 1
        else
            # These are different characters, so the only option is that second < first
            echo 1
            return 2
        end
    end
    # No components differ, version is the same.
    echo 0
    return 0
end
```

On the other hand, it seems to give the same results as `vercmp` on anything I've tried, and it only uses builtins, so it should be system-independent.

**Comment by abudden** [comment]
Thanks for all the great suggestions.  I think that for now I'll use the `sort -CV` suggestion as I (currently) don't use Mac OS X and I've one server I use has Ubuntu 16.04 LTS with Fish 2.2.0, so the string stuff isn't supported.

In the longer term I'd certainly expect to use Mac OS X at some point, so it would be nice to have a platform agnostic approach (although by that time I'd expect that I'd be looking at version differences between 3.x and maybe even 4.x!).

So, my ideal answer to this would be to have two things:

1. A built-in method of checking that `$FISH_VERSION` is greater than or equal to a particular version (supported from version 3.0 onward).
2. A documented recommended way of using it that will return false **silently** if the version is less than the one that added support for version checking - I haven't found a way to test for the existence of a function or suppress the error message that comes from calling an unknown function, so I don't know the best way to do this.

One option for this might be something like:

```fish
if test (echo $FISH_VERSION | cut -d . -f 1) -ge 3 ; and fish_vercheck 3.2.1
    # Do something that needs 3.2.1.
end
```

Of course others may be able to suggest something more concise than that.

**Comment by floam** [comment]
I'm on MacOS (High Sierra) and `sort -V` seems to work here.

```console
$ echo 1.5\n1.2a\n1.2.0\n1.4 | /usr/bin/sort -V
1.2a
1.2.0
1.4
1.5
```

**Comment by faho** [comment]
>    if test (echo $FISH_VERSION | cut -d . -f 1) -ge 3 ; and fish_vercheck 3.2.1

Any `fish_vercheck` implementation should be able to compare versions without that first check. So it'll just be `if fish_vercheck 3.2.1`. Which should exclude "2.6", "3.1" and "3.2.0", but include "3.3.0" and "3.2.1-12-gSOMEHASH".

**Comment by abudden** [comment]
> > if test (echo $FISH_VERSION | cut -d . -f 1) -ge 3 ; and fish_vercheck 3.2.1

> Any fish_vercheck implementation should be able to compare versions without that first check. So it'll > just be if fish_vercheck 3.2.1. Which should exclude "2.6", "3.1" and "3.2.0", but include "3.3.0" and "3.2.1-12-gSOMEHASH".

I agree that this would be nice, but if it's a built-in function (which is what I was suggesting would be beneficial), then on Fish 2.2.0, you'd get:

```
[I] al@HOSTNAME ~> if fish_vercheck 3.2.1
                       echo "Okay version"
                   end
fish: Unknown command 'fish_vercheck'
fish:
if fish_vercheck 3.2.1
   ^
[I] al@HOSTNAME ~> # Compare the above to:
[I] al@HOSTNAME ~> if test (echo $FISH_VERSION | cut -d . -f 1) -ge 3 ; and fish_vercheck 3.2.1
                       echo "Okay version"
                   end
[I] al@HOSTNAME ~> 
```

and I was trying to come up with something that would both work out-of-the-box and not produce an error message when the test fails.

**Comment by abudden** [comment]
Having read the manual a bit more, with a 3.0.0-supported built-in, my test line (from [here](https://github.com/fish-shell/fish-shell/issues/4419#issuecomment-331383183)) should probably look like this:

```fish
if functions -q fish_vercheck; and fish_vercheck 3.2.1
	# Do something that needs 3.2.1.
end
```

For now, I've implemented a GNU sort-based version check as follows:

```fish
# # Example of how to test for specific versions
# # using my sort-based implementation:
#
# if testversion 2.4.0
# end
#
# # To test against a different version (other than Fish's version):
# # Note that different applications will require different mangling
# # to get the actual version.
#
# if testversion 2.7.0 (python --version 2>&1 | cut -d \  -f 2)
# end
function testversion -a expected actual -d "Compare versions.  By default this tests \$FISH_VERSION, but if a second argument is provided it tests against that version."
	if test -z $actual
		set actual $FISH_VERSION
	end
	begin ; echo $expected ; echo $actual ; end | sort --check=silent --version-sort
	return $status
end
```

**Comment by faho** [comment]
@abudden:

Never use `test` without quotes - it's one of the few parts of fish that is completely POSIX-compatible, so it also has some of POSIX's wrong decisions. `test -z` without further arguments is true, so `test -z $var` is also true if $var is _empty_. I.e. `testversion 2.6.0` is true, no matter what the actual version is. You'd need to call `testversion 2.6.0 ""` to get it to default to $FISH_VERSION.

Either use `test -z "$actual"` or `not set -q actual[1]` (i.e. actual has no first element).

Also, the `begin; echo $expected; echo $actual; end` can be replaced with `printf '%s\n' $expected $actual`. Personally I think that's a bit simpler.

>if functions -q fish_vercheck; and fish_vercheck 3.2.1

Basically. Though I'd use `type -q`, which is also true if it's a builtin or command.

**Comment by abudden** [comment]
@faho:

Great, thanks for the advice.  I've amended my version:

```fish
# # Example of how to test for specific versions
# # using my sort-based implementation:
#
# if testversion 2.4.0
# end
#
# # To test against a different version (other than Fish's version):
# # Note that different applications will require different mangling
# # to get the actual version.
#
# if testversion 2.7.0 (python --version 2>&1 | cut -d \  -f 2)
# end
function testversion -a expected actual -d "Compare versions.  By default this tests \$FISH_VERSION, but if a second argument is provided it tests against that version."
	if test -z "$actual"
		set actual $FISH_VERSION
	end
	printf '%s\n' $expected $actual | sort --check=silent --version-sort
	return $status
end
```

**Comment by edouard-lopez** [comment]
To run on a **docker container, _i.e._ busybox****, you can rely neither on `--check=silent` nor `--version-sort`, but [you can achieve it with `-c` and several `-k`](https://stackoverflow.com/a/4495368/802365):

### First is lower than second
```fish
❯ echo -e '1.0.0\n3.0.0' | sort --check=silent --version-sort; echo $status
0
❯ echo -e '1.0.0\n3.0.0' | sort -c -t. -k 1,1n -k 2,2n -k 3,3n  2>&1 /dev/null; echo $status
0
```

### First is greater than second


```fish
❯ echo -e '3.0.0\n1.0.0' | sort --check=silent --version-sort; echo $status
1
❯ echo -e '3.0.0\n1.0.0' | sort -c -t. -k 1,1n -k 2,2n -k 3,3n 2> /dev/null; echo $status
1 
```

**Comment by zanchey** [comment]
That will blow up with 3.0b1, for example.
